src/client-side-metrics/metrics-tracer-factory.ts

```diff
@@ -0,0 +1,600 @@
+    return Object.assign(
+      {
+        attemptStatus: attemptStatus,
+      },
+      this.getBasicAttributes(projectId)
+    );
+  }
+
+  /**
+   * Called when the operation starts. Records the start time.
+   */
+  onOperationStart() {
+    this.operationStartTime = this.dateProvider.getDate();
+  }
+
+  /**
+   * Called after the client reads a row. Records application blocking latencies.
+   */
+  onRead() {
+    const currentTime = this.dateProvider.getDate();
+    const projectId = this.projectId;
+    if (this.lastReadTime) {
+      if (projectId && this.lastReadTime) {
+        const attributes = this.getBasicAttributes(projectId);
+        const difference = currentTime.getTime() - this.lastReadTime.getTime();
+        this.metrics.applicationBlockingLatencies.record(
+          difference,
+          attributes
+        );
+        this.lastReadTime = currentTime;
+      }
+    } else {
+      this.lastReadTime = currentTime;
+    }
+  }
+
+  /**
+   * Called when an attempt (e.g., an RPC attempt) completes. Records attempt latencies.
+   * @param {AttemptInfo} info Information about the completed attempt.
+   */
+  onAttemptComplete(info: AttemptInfo) {
+    const endTime = this.dateProvider.getDate();
+    const projectId = this.projectId;
+    if (projectId && this.attemptStartTime) {
+      const attributes = this.getAttemptAttributes(
+        projectId,
+        info.finalOperationStatus,
+        info.streamingOperation
+      );
+      const totalTime = endTime.getTime() - this.attemptStartTime.getTime();
+      this.metrics.attemptLatencies.record(totalTime, attributes);
+    }
+  }
+
+  /**
+   * Called when a new attempt starts. Records the start time of the attempt.
+   */
+  onAttemptStart() {
+    this.attemptStartTime = this.dateProvider.getDate();
+  }
+
+  /**
+   * Called when the first response is received. Records first response latencies.
+   * @param {string} finalOperationStatus The final status of the operation.
+   */
+  onResponse(finalOperationStatus: string) {
+    const endTime = this.dateProvider.getDate();
+    const projectId = this.projectId;
+    if (projectId && this.operationStartTime) {
+      const attributes = this.getFinalOpAttributes(
+        projectId,
+        finalOperationStatus
+      );
+      const totalTime = endTime.getTime() - this.operationStartTime.getTime();
+      if (!this.receivedFirstResponse) {
+        this.receivedFirstResponse = true;
+        this.metrics.firstResponseLatencies.record(totalTime, attributes);
+      }
+    }
+  }
+
+  /**
+   * Called when an operation completes (successfully or unsuccessfully).
+   * Records operation latencies, retry counts, and connectivity error counts.
+   * @param {OperationInfo} info Information about the completed operation.
+   */
+  onOperationComplete(info: OperationInfo) {
+    const endTime = this.dateProvider.getDate();
+    const projectId = this.projectId;
+    this.onAttemptComplete(info);
```

Comment:
another comment about adding some kind of state machine. It looks like there are no guards against calling onAttemptComplete or onOperationComplete multiple times, which could pollute the data

Reply:
Upon further consideration I have decided to eliminate this onAttemptComplete call. This will require one more line of code from the method collecting the metrics, but I think requiring the method to be more explicit when collecting metrics is a better move.